### PR TITLE
Add clang-iwyu

### PR DESCRIPTION
**Pull request by mehrdadn**
## Why are these changes needed?

Checks that headers are included correctly, to prevent e.g. cross-platform errors (and for general C++ hygiene).

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: src/ray/protobuf/extra_actions_base.proto [hunk 1]
```diff
@@ -0,0 +1,184 @@
+// Copyright 2014 The Bazel Authors. All rights reserved.
```
Comment (simon-mo) [inline src/ray/protobuf/extra_actions_base.proto]: Does this have to be in our `src`?

**Review by simon-mo — COMMENTED** [review]
Does it run on windows, mac, and linux

File: src/ray/protobuf/extra_actions_base.proto (hunk 1 above)
Comment (mehrdadn) [reply src/ray/protobuf/extra_actions_base.proto]: Until this issue is resolved: https://github.com/mehrdadn/ray/blob/include-what-you-use/BUILD.bazel#L846

Comment (mehrdadn) [reply src/ray/protobuf/extra_actions_base.proto]: Not sure about Mac, but the script does run on Windows and Linux.  
The CI only runs for Linux though; I think that's enough to catch most issues.

Comment (simon-mo) [reply src/ray/protobuf/extra_actions_base.proto]: Ok however can we move it to thirdparty, patches, or other directory so we don't confuse this proto with ray's protos?

Comment (mehrdadn) [reply src/ray/protobuf/extra_actions_base.proto]: Sure!

**Comment by simon-mo** [comment]
Also, would you recommend merge this build step and add a separate PR to make it pass, or use this PR to fix all the iwyu warnings?

**Comment by mehrdadn** [comment]
That should definitely be a separate PR, it's a significant undertaking and not likely to happen for a while.

**Comment by simon-mo** [comment]
In that case I would like to hold off from merging this so our build won’t be failing all the time.

**Comment by mehrdadn** [comment]
But then people can't actually run the code to fix the problems?

We can address your concern easily though, either by disabling the build or via https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#jobsjob_idstepscontinue-on-error

**Review by simon-mo — APPROVED** [review]

**Comment by simon-mo** [comment]
Bazel lint failed https://travis-ci.com/ray-project/ray/jobs/284729480#L379

```
$ ./ci/travis/bazel-format.sh
~/build/ray-project/ray ~/build/ray-project/ray
BUILD.bazel:
--- BUILD.bazel	2020-02-07 05:39:38.652255300 +0000
+++ /tmp/buildifier-tmp-138618377	2020-02-07 05:53:46.036414728 +0000
@@ -852,21 +852,20 @@
 
 action_listener(
     name = "iwyu_cpp",
-    mnemonics = ["CppCompile"],
     extra_actions = [":iwyu_action"],
+    mnemonics = ["CppCompile"],
 )
 
 extra_action(
     name = "iwyu_action",
+    cmd = "$(location :iwyu_sh) $(location @com_google_protobuf//:protoc) $(location :extra_actions_base_proto) --extra_action_file=$(EXTRA_ACTION_FILE)",
     tools = [
-        ":iwyu_sh",
         ":extra_actions_base_proto",
+        ":iwyu_sh",
         "@com_google_protobuf//:protoc",
     ],
-    cmd = "$(location :iwyu_sh) $(location @com_google_protobuf//:protoc) $(location :extra_actions_base_proto) --extra_action_file=$(EXTRA_ACTION_FILE)",
 )
 
-
 cc_library(
     name = "sha256",
     srcs = [
exit status 1
The command "./ci/travis/bazel-format.sh" exited with 4.
```

**Comment by mehrdadn** [comment]
Oops, thanks! Just pushed a fix.

**Comment by simon-mo** [comment]
Lint passed. merging

**Pull request merged**